drivers/led/lp50xx.c

```diff
@@ -0,0 +1,475 @@
+	int err;
+
+	/* Software reset */
+	buf[0] = config->reset_reg;
+	buf[1] = RESET_SW;
+	err = i2c_write_dt(&config->bus, buf, 2);
+	if (err < 0) {
+		return err;
+	}
+
+	/* After reset, apply configuration since all registers are reset. */
+	buf[0] = LP50XX_DEVICE_CONFIG1;
+	buf[1] = CONFIG1_PWM_DITHERING_EN | CONFIG1_AUTO_INCR_EN
+		| CONFIG1_POWER_SAVE_EN;
+	if (config->max_curr_opt) {
+		buf[1] |= CONFIG1_MAX_CURRENT_OPT;
+	}
+	if (config->log_scale_en) {
+		buf[1] |= CONFIG1_LOG_SCALE_EN;
+	}
+
+	return i2c_write_dt(&config->bus, buf, 2);
+}
+
+static int lp50xx_hw_enable(const struct device *dev, bool enable)
+{
+	const struct lp50xx_config *config = dev->config;
+	int err;
+
+	if (config->gpio_enable.port == NULL) {
+		/* Nothing to do */
+		return 0;
+	}
+
+	err = gpio_pin_set_dt(&config->gpio_enable, enable);
+	if (err < 0) {
+		LOG_ERR("%s: failed to set enable gpio", dev->name);
+		return err;
+	}
+
+	k_usleep(LP50XX_ENABLE_DELAY_US);
+
+	return 0;
+}
+
+static int lp50xx_enable(const struct device *dev, bool enable)
+{
+	const struct lp50xx_config *config = dev->config;
+	uint8_t value = enable ? CONFIG0_CHIP_EN : 0;
+
+	return i2c_reg_update_byte_dt(&config->bus,
+				      LP50XX_DEVICE_CONFIG0,
+				      CONFIG0_CHIP_EN,
+				      value);
+}
+
+static int lp50xx_init(const struct device *dev)
+{
+	const struct lp50xx_config *config = dev->config;
+	int err;
+
+	if (!device_is_ready(config->bus.bus)) {
+		LOG_ERR("%s: I2C device not ready", dev->name);
+		return -ENODEV;
+	}
+
+	if (config->num_leds > config->max_leds) {
+		LOG_ERR("%s: invalid number of LEDs %d (max %d)",
+			dev->name,
+			config->num_leds,
+			config->max_leds);
+		return -EINVAL;
+	}
+
+	/* Configure GPIO if present */
+	if (config->gpio_enable.port != NULL) {
+		if (!device_is_ready(config->gpio_enable.port)) {
+			LOG_ERR("%s: enable gpio is not ready", dev->name);
+			return -ENODEV;
+		}
+
+		err = gpio_pin_configure_dt(&config->gpio_enable,
+					    GPIO_OUTPUT_INACTIVE);
+		if (err < 0) {
+			LOG_ERR("%s: failed to initialize enable gpio",
+				dev->name);
+			return err;
+		}
+	}
+
+	/* Enable hardware */
+	err = lp50xx_hw_enable(dev, true);
+	if (err < 0) {
+		LOG_ERR("%s: failed to enable hardware", dev->name);
+		return err;
+	}
+
+	/* Reset device */
+	err = lp50xx_reset(dev);
+	if (err < 0) {
+		LOG_ERR("%s: failed to reset", dev->name);
+		return err;
+	}
+
+	/* Enable device */
+	err = lp50xx_enable(dev, true);
+	if (err < 0) {
+		LOG_ERR("%s: failed to enable", dev->name);
+		return err;
+	}
+
+	return 0;
+}
+
+#ifdef CONFIG_PM_DEVICE
+static int lp50xx_pm_action(const struct device *dev,
+			    enum pm_device_action action)
+{
+	int err;
+
+	switch (action) {
+	case PM_DEVICE_ACTION_SUSPEND:
+		return lp50xx_enable(dev, false);
+	case PM_DEVICE_ACTION_RESUME:
+		return lp50xx_enable(dev, true);
+	case PM_DEVICE_ACTION_TURN_OFF:
+		return lp50xx_hw_enable(dev, false);
+	case PM_DEVICE_ACTION_TURN_ON:
+		err = lp50xx_hw_enable(dev, true);
```

Comment:
this is reserved for power domains

Reply:
Thanks for your help @gmarull.

I do have a question for you:

On some systems we may need to disable PM support for a LED device/driver. For example, a LED can be used to display a pattern when the device runs in a low-power mode. In your opinion, how could we achieve this ?

Reply:
OK, I've read the documentation through and through and I don't understand why this implementation could not work (except for the backup of registers to restore configuration). Why wouldn't you add a LED controller to a power domain ?

For your question, is there anything like a "software" power domain ? The LED could be added to one and therefore the application can be responsible for turning it on and off. Since I am new to Zephyr, it may be a silly idea but I think it is the kind of thing that could work.

For the original discussion at hand, I will remove the `TURN_ON`/`TURN_OFF` actions because, as you said, it won't work as expected and needs more work.

Reply:
@man-gc Note that you can add "gpio enable" support to the `SUSPEND` / `RESUME` PM action handling. But you'll need to restore the configuration. As an alternative, you can give "gpio enable" control to apps via a private API (like in the [is31fl3733 driver](https://github.com/zephyrproject-rtos/zephyr/pull/56673/commits/45605a05e4ec8770793d1ab1fd7ae8b8b036056a).